engine/runtime/src/main/java/org/enso/interpreter/node/controlflow/caseexpr/DateBranchNode.java

```diff
@@ -0,0 +1,59 @@
+package org.enso.interpreter.node.controlflow.caseexpr;
+
+import com.oracle.truffle.api.RootCallTarget;
+import com.oracle.truffle.api.dsl.Fallback;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.api.interop.InteropLibrary;
+import com.oracle.truffle.api.library.CachedLibrary;
+import com.oracle.truffle.api.nodes.NodeInfo;
+import com.oracle.truffle.api.profiles.ConditionProfile;
+import org.enso.interpreter.runtime.data.Type;
+import org.enso.interpreter.runtime.data.EnsoDate;
+
+/** An implementation of the case expression specialised to working on Date. */
+@NodeInfo(shortName = "DateMatch")
+public abstract class DateBranchNode extends BranchNode {
+  private final Type date;
+  private final ConditionProfile profile = ConditionProfile.createCountingProfile();
+
+  DateBranchNode(Type vector, RootCallTarget branch) {
+    super(branch);
+    this.date = vector;
+  }
+
+  /**
+   * Creates a new node for handling matching on a case expression.
+   *
+   * @param date the expression to use for matching
+   * @param branch the expression to be executed if (@code matcher} matches
+   * @return a node for matching in a case expression
+   */
+  public static DateBranchNode build(Type date, RootCallTarget branch) {
+    return DateBranchNodeGen.create(date, branch);
+  }
+
+  @Specialization
+  void doType(VirtualFrame frame, Object state, Type target) {
+    if (profile.profile(date == target)) {
+      accept(frame, state, new Object[0]);
+    }
+  }
+
+  @Specialization
+  void doEnsoDate(VirtualFrame frame, Object state, EnsoDate date) {
+    accept(frame, state, new Object[0]);
+  }
+
+  @Specialization(guards = "interop.isDate(date)")
+  void doDate(
+      VirtualFrame frame,
+      Object state,
+      Object date,
+      @CachedLibrary(limit = "10") InteropLibrary interop) {
+    accept(frame, state, new Object[0]);
+  }
```

Comment:
If we don't check `!interop.isTime(dateTime)` then a `Date_Time` will also match a `Date` branch.

This may make sense, but it's not consistent with our type system where we don't have `Date_Time` as a subtype of `Date`. And in most cases I think we just want to distinguish the two as separate types - `Date` in Enso represents a concept of a date without time, not a date with maybe time.

Reply:
Similar argument can be made about `Vector` and `Array` and I don't think the answer is so straightforward. See https://www.pivotaltracker.com/story/show/183266964/comments/233361793. 
We should probably take it offline